tests/admin_views/test_autocomplete_view.py

```diff
@@ -0,0 +1,173 @@
+from __future__ import unicode_literals
+
+import json
+import sys
+
+from django.contrib.admin import site
+from django.contrib.admin.tests import AdminSeleniumTestCase
+from django.contrib.admin.views.autocomplete import (
+    AutocompleteJsonView, AutocompletePage, AutocompletePaginator,
+)
+from django.contrib.auth.models import Permission, User
+from django.core.exceptions import PermissionDenied
+from django.http import Http404
+from django.test import RequestFactory, TestCase, override_settings
+from django.urls import reverse, reverse_lazy
+from django.utils.six import text_type
+
+from .admin import AnswerAdmin, QuestionAdmin
+from .models import Answer, Question
+from .tests import AdminViewBasicTestCase
+
+site.register(Question, QuestionAdmin)
+site.register(Answer, AnswerAdmin)
+
+
+class AutocompleteJsonViewTest(AdminViewBasicTestCase):
+    url = reverse_lazy('admin:autocomplete')
+
+    @classmethod
+    def setUpTestData(cls):
+        cls.user = User.objects.create_user(username='user', password='secret', email='user@example.com')
+        AdminViewBasicTestCase.setUpTestData()
+
+    def test_no_field_identifier(self):
+        factory = RequestFactory()
+        request = factory.get(self.url)
+        request.user = self.superuser
+        with self.assertRaisesMessage(PermissionDenied, 'No "field_identifier" provided.'):
+            AutocompleteJsonView.as_view(admin_site=site)(request)
+
+    def test_wrong_field_identifier_format(self):
+        factory = RequestFactory()
+        request = factory.get(self.url, {'field_identifier': 'wrong.format'})
+        request.user = self.superuser
+        with self.assertRaisesMessage(PermissionDenied, 'Invalid "field_identifier".'):
+            AutocompleteJsonView.as_view(admin_site=site)(request)
+
+    def test_not_existing_field_identifier(self):
+        factory = RequestFactory()
+        request = factory.get(self.url, {'field_identifier': 'foo.bar.baz'})
+        request.user = self.superuser
+        with self.assertRaisesMessage(PermissionDenied, 'Invalid "field_identifier".'):
+            AutocompleteJsonView.as_view(admin_site=site)(request)
+
+    def test_not_model_not_registered(self):
+        site.unregister(Question)
+        factory = RequestFactory()
+        request = factory.get(self.url, {'field_identifier': 'admin_views.Answer.question'})
+        request.user = self.superuser
+        with self.assertRaisesMessage(
```

Comment:
To avoid the interesting indentation:
```
msg = "<class 'admin_views.models.Question'> is not registered in the admin."
with self.assertRaisesMessage(Http404, msg):
```